e2e/process_test.go

```diff
@@ -0,0 +1,469 @@
+	for i := 0; i < cfg.clusterSize; i++ {
+		port := etcdProcessBasePort + 2*i
+		curl := url.URL{Scheme: clientScheme, Host: fmt.Sprintf("localhost:%d", port)}
+		purl := url.URL{Scheme: peerScheme, Host: fmt.Sprintf("localhost:%d", port+1)}
+		name := fmt.Sprintf("testname%d", i)
+		dataDirPath := name + ".etcd"
+		initialCluster[i] = fmt.Sprintf("%s=%s", name, purl.String())
+
+		args := []string{
+			"--name", name,
+			"--listen-client-urls", curl.String(),
+			"--advertise-client-urls", curl.String(),
+			"--listen-peer-urls", purl.String(),
+			"--initial-advertise-peer-urls", purl.String(),
+			"--initial-cluster-token", cfg.initialToken,
+			"--data-dir", dataDirPath,
+		}
+		if cfg.isClientTLS {
+			tlsClientArgs := []string{
+				"--cert-file", certPath,
+				"--key-file", privateKeyPath,
+				"--ca-file", caPath,
+			}
+			args = append(args, tlsClientArgs...)
+		}
+		if cfg.isPeerTLS {
+			tlsPeerArgs := []string{
+				"--peer-cert-file", certPath,
+				"--peer-key-file", privateKeyPath,
+				"--peer-ca-file", caPath,
+			}
+			args = append(args, tlsPeerArgs...)
+		}
+
+		etcdCfgs[i] = &etcdProcessConfig{
+			args:        args,
+			name:        name,
+			dataDirPath: dataDirPath,
+			purl:        purl,
+			acurl:       curl,
+		}
+	}
+	for i := 0; i < cfg.proxySize; i++ {
+		port := etcdProcessBasePort + 2*cfg.clusterSize + i + 1
+		curl := url.URL{Scheme: clientScheme, Host: fmt.Sprintf("localhost:%d", port)}
+		name := fmt.Sprintf("testname-proxy%d", i)
+		dataDirPath := name + ".etcd"
+		args := []string{
+			"--name", name,
+			"--proxy", "on",
+			"--listen-client-urls", curl.String(),
+			"--data-dir", dataDirPath,
+		}
+		etcdCfgs[cfg.clusterSize+i] = &etcdProcessConfig{
+			args:        args,
+			name:        name,
+			dataDirPath: dataDirPath,
+			acurl:       curl,
+			isProxy:     true,
+		}
+	}
+
+	initialClusterArgs := []string{"--initial-cluster", strings.Join(initialCluster, ",")}
+	for i := range etcdCfgs {
+		etcdCfgs[i].args = append(etcdCfgs[i].args, initialClusterArgs...)
+	}
+
+	return etcdCfgs
+}
+
+func (epc *etcdProcessCluster) Close() (err error) {
+	for _, p := range epc.procs {
+		if p == nil {
+			continue
+		}
+		os.RemoveAll(p.cfg.dataDirPath)
+		if curErr := p.proc.Close(); curErr != nil {
+			if err != nil {
+				err = fmt.Errorf("%v; %v", err, curErr)
+			} else {
+				err = curErr
+			}
+		}
+		<-p.donec
+	}
+	return err
+}
+
+func spawnCmd(args []string) (*gexpect.ExpectSubprocess, error) {
+	// redirect stderr to stdout since gexpect only uses stdout
+	cmdArgs := `/bin/sh -c "` + strings.Join(args, " ") + ` 2>&1 "`
+	return gexpect.Spawn(cmdArgs)
+}
+
+func spawnWithExpect(args []string, expected string) error {
+	proc, err := spawnCmd(args)
+	if err != nil {
+		return err
+	}
+	ok, err := proc.ExpectRegex(expected)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		return fmt.Errorf("couldn't get expected output: '%s'", expected)
+	}
+	return proc.Close()
```

Comment:
Nevermind the defer bit-- Close can't be deferred because the success case needs the close error code. The proc still needs a close on the error paths, though.